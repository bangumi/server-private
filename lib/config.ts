/**
 * 从 `projectRoot/config.yaml` 和环境变量读取配置。 配置文件见 `config.example.yaml`。
 *
 * {@link schema} 定义了配置文件的 json schema
 *
 * 可以使用环境变量会覆盖对应的文件配置，如 `MYSQL_DB` 会覆盖 `mysql.db`, 具体可用的环境变量和对应的配置项也定义在 {@link schema}。
 *
 * 除此之外可以设置的环境变量：
 *
 * {@link NODE_ENV}
 *
 * 'production' | 'stage' | 'test' 在测试时会被 vitest 会设置此环境变量为 'test'， 在生产环境会被设置为 'production',
 * 在测试部属环境被会设置为 stage
 *
 * {@link HTTPS_PROXY}
 *
 * 默认为空，如果设置了的话，会作为 turnstile 的代理
 *
 * {@link REF}
 *
 * 用于 docker 镜像，不用设置。
 *
 * @packageDocumentation
 */
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as process from 'node:process';
import * as url from 'node:url';

import type { Static, TSchema } from '@sinclair/typebox';
import { Kind, Type as t } from '@sinclair/typebox';
import { Value } from '@sinclair/typebox/value';
import Ajv from 'ajv';
import addFormats from 'ajv-formats';
import * as yaml from 'js-yaml';
import * as lo from 'lodash-es';

// read from env

const { HTTPS_PROXY = '', NODE_ENV, REF } = process.env;

export const production = NODE_ENV === 'production';
export const stage = NODE_ENV === 'stage';
export const developing = NODE_ENV === 'development';
export const testing = NODE_ENV === 'test';

export const projectRoot = url.fileURLToPath(new URL('..', import.meta.url));
export const pkg = JSON.parse(
  fs.readFileSync(path.resolve(projectRoot, 'package.json'), 'utf8'),
) as { version: string };

export const VERSION = developing ? 'development' : REF || pkg.version;

export const redisPrefix = `graphql-${VERSION}`;

export { HTTPS_PROXY };

/** WARNING: 所有的 key 都必需小写，而且不能包含下划线。 */
const schema = t.Object({
  nsfw_word: t.Optional(t.String({ minLength: 1 })),
  disable_words: t.Optional(t.String()),
  banned_domain: t.Optional(t.String()),

  redisUri: t.String({ default: 'redis://127.0.0.1:3306/0', env: 'REDIS_URI' }),

  image: t.Object({
    provider: t.Enum(
      {
        s3: 's3',
        FS: 'fs',
      } as const,
      { default: 'fs', env: 'CHII_IMAGE_PROVIDER' },
    ),
    imaginaryUrl: t.Optional(
      t.String({
        description: 'url to docker image running https://github.com/h2non/imaginary',
      }),
    ),
    fs: t.Object({
      path: t.String({ default: './tmp/images' }),
    }),
    s3: t.ReadonlyOptional(
      t.Object({
        endPoint: t.String({ env: 'CHII_IMAGE_S3_ENDPOINT' }),
        bucket: t.String({ default: 'chii-image', env: 'CHII_IMAGE_S3_BUCKET' }),
        port: t.Integer({ default: 9000, env: 'CHII_IMAGE_S3_PORT' }),
        useSSL: t.Boolean({ default: false, env: 'CHII_IMAGE_S3_USE_SSL' }),
        accessKey: t.String({ env: 'CHII_IMAGE_S3_ACCESS_KEY' }),
        secretKey: t.String({ env: 'CHII_IMAGE_S3_SECRET_KEY' }),
      }),
    ),
  }),

  turnstile: t.Object({
    secretKey: t.String({
      default: '1x0000000000000000000000000000000AA',
      env: 'TURNSTILE_SECRET_KEY',
    }),
    siteKey: t.String({ default: '1x00000000000000000000AA', env: 'TURNSTILE_SITE_KEY' }),
  }),

  mysql: t.Object({
    db: t.String({ default: 'bangumi', env: 'MYSQL_DB' }),
    host: t.String({ default: '127.0.0.1', env: 'MYSQL_HOST' }),
    port: t.Integer({ default: 3306, env: 'MYSQL_PORT' }),
    user: t.String({ default: 'user', env: 'MYSQL_USER' }),
    password: t.String({ default: 'password', env: 'MYSQL_PASS' }),
  }),
});

// read config file

export default readConfig();

function readConfig(): Static<typeof schema> {
  const configFilePath = path.resolve(projectRoot, 'config.yaml');

  let configFileContent = '{}';
  if (fs.existsSync(configFilePath)) {
    configFileContent = fs.readFileSync(configFilePath, 'utf8');
  }

  const config = lo.assign(Value.Create(schema), yaml.load(configFileContent));

  function readFromEnv(keyPath: string[], o: TSchema) {
    if (o[Kind] === 'Object') {
      for (const [key, value] of Object.entries(o.properties as Record<string, TSchema>)) {
        readFromEnv([...keyPath, key], value);
      }

      return;
    }

    const envKey = o.env as string | undefined;
    if (envKey) {
      const v = process.env[envKey];
      if (v) {
        lo.set(config, keyPath, process.env[envKey]);
      }
    }
  }

  readFromEnv([], schema);

  const ajv = new Ajv({ allErrors: true, coerceTypes: true, keywords: ['env'] });
  addFormats(ajv);

  const check = ajv.compile(schema);

  const valid = check(config);

  if (!valid) {
    const errorMessage =
      check.errors
        ?.map(
          (x) => '  ' + (x.instancePath + ': ' + (x.message ?? `wrong data type ${x.schemaPath}`)),
        )
        .join('\n') ?? '';

    throw new TypeError('failed to validate config file:\n' + errorMessage);
  }

  return config;
}
