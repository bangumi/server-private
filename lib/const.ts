export const TagCat = {
  subject: 0,
  meta: 3,
} as const;
