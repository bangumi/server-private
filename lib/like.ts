import * as lo from 'lodash-es';

import { db, op, schema } from '@app/drizzle';
import * as fetcher from '@app/lib/types/fetcher.ts';
import type * as res from '@app/lib/types/res.ts';

export enum LikeType {
  SubjectCover = 1,

  GroupTopic = 7,
  GroupReply = 8,

  SubjectReply = 10,

  EpReply = 11,

  SubjectCollect = 40,
}

export const LIKE_REACTIONS_ALLOWED: ReadonlySet<number> = Object.freeze(
  new Set([
    0, // bgm67
    140, // bgm124
    80, // bgm64
    54, // bgm38
    85, // bgm69

    104, // bgm88
    88, // bgm72
    62, // bgm46
    79, // bgm63
    53, // bgm37

    122, // bgm106
    92, // bgm76
    118, // bgm102
    141, // bgm125
    90, // bgm74

    76, // bgm60
    60, // bgm44
    128, // bgm112
    47, // bgm31
    68, // bgm52

    137, // bgm121
    132, // bgm116
  ]),
);

export async function fetchTopicReactions(
  topicID: number,
  type: LikeType,
): Promise<Record<number, res.IReaction[]>> {
  const data = await db
    .select()
    .from(schema.chiiLikes)
    .where(op.and(op.eq(schema.chiiLikes.mainID, topicID), op.eq(schema.chiiLikes.type, type)));

  const uids = data.map((x) => x.uid);
  const users = await fetcher.fetchSimpleUsersByIDs(uids);
  const r = lo.groupBy(data, (x) => x.relatedID);

  return lo.mapValues(r, (v): res.IReaction[] => {
    return Object.entries(
      lo.groupBy(v, (a) => {
        return a.value;
      }),
    ).map(([key, values]) => {
      return {
        users: values.map((x) => users[x.uid]).filter((x) => x !== undefined),
        value: Number(key),
      };
    });
  });
}

export async function fetchSubjectCollectReactions(
  collectIDs: number[],
): Promise<Record<number, res.IReaction[]>> {
  const data = await db
    .select()
    .from(schema.chiiLikes)
    .where(
      op.and(
        op.eq(schema.chiiLikes.type, LikeType.SubjectCollect),
        op.inArray(schema.chiiLikes.relatedID, collectIDs),
      ),
    );

  const uids = data.map((x) => x.uid);
  const users = await fetcher.fetchSimpleUsersByIDs(uids);
  const r = lo.groupBy(data, (x) => x.relatedID);

  return lo.mapValues(r, (v): res.IReaction[] => {
    return Object.entries(
      lo.groupBy(v, (a) => {
        return a.value;
      }),
    ).map(([key, values]) => {
      return {
        users: values.map((x) => users[x.uid]).filter((x) => x !== undefined),
        value: Number(key),
      };
    });
  });
}
