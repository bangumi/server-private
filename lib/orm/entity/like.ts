import { DateTime } from 'luxon';
import { Column, Entity, Index } from 'typeorm';

/** 用于点赞/封面投票 */
@Index('idx_uid', ['uid'], {})
@Index('idx_related', ['relatedID'], {})
@Entity('chii_likes', { schema: 'bangumi' })
export class Like {
  static readonly TYPE_SUBJECT_COVER = 1;

  @Column('mediumint', { primary: true, name: 'type', unsigned: true })
  type!: number;

  @Column('mediumint', { primary: true, name: 'related_id' })
  relatedID!: number;

  @Column('mediumint', { primary: true, name: 'uid', unsigned: true })
  uid!: number;

  @Column('mediumint', { name: 'value', unsigned: true })
  value = 0;

  @Column('tinyint', { name: 'ban', unsigned: true, default: () => "'0'" })
  ban = 0;

  /** Use `@CreateDateColumn` after https://github.com/typeorm/typeorm/issues/8701 is fixed */
  @Column('int', {
    name: 'created_at',
    comment: 'unix timestamp seconds',
    unsigned: true,
    transformer: {
      to: (value: Date) => Math.trunc(value.getTime() / 1000),
      from: (value: number) => DateTime.fromSeconds(value).toJSDate(),
    },
  })
  createdAt!: Date;
}
