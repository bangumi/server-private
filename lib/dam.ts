/** 敏感词过滤 */
import config from './config.ts';

const controlPattern = /[\u200B-\u200F\u202A-\u202E\uFEFF]/;

interface Option {
  nsfw_word?: string;
  disabled_words?: string;
  banned_domain?: string;
}

export class Dam {
  private readonly nsfwWord: RegExp | undefined;
  private readonly disableWord: RegExp | undefined;
  private readonly bannedDomain: RegExp | undefined;

  constructor({ banned_domain, disabled_words, nsfw_word }: Option) {
    if (nsfw_word) {
      this.nsfwWord = new RegExp(nsfw_word, 'i');
    }

    if (disabled_words) {
      this.disableWord = new RegExp(disabled_words, 'i');
    }

    if (banned_domain) {
      this.bannedDomain = new RegExp(banned_domain, 'i');
    }
  }

  needReview(text: string): boolean {
    if (text === '') {
      return false;
    }

    if (!this.disableWord) {
      return false;
    }

    return this.disableWord.test(text);
  }

  censoredWords(text: string): boolean {
    if (this.disableWord?.test(text)) {
      return true;
    }

    if (this.bannedDomain != undefined) {
      return this.bannedDomain.test(text);
    }

    return false;
  }

  static allCharacterPrintable(s: string): boolean {
    return !controlPattern.test(s);
  }
}

export const dam = new Dam(config);
