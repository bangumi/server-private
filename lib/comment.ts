import { DateTime } from 'luxon';

import { db, op } from '@app/drizzle/db';
import * as schema from '@app/drizzle/schema';
import type { IAuth } from '@app/lib/auth/index.ts';
import { NotAllowedError } from '@app/lib/auth/index.ts';
import { Dam } from '@app/lib/dam.ts';
import { BadRequestError, NotFoundError } from '@app/lib/error.ts';
import { fetchTopicReactions, LikeType } from '@app/lib/like.ts';
import { CommentState } from '@app/lib/topic/type.ts';
import * as fetcher from '@app/lib/types/fetcher.ts';
import type * as res from '@app/lib/types/res.ts';
import { LimitAction } from '@app/lib/utils/rate-limit/index.ts';
import { rateLimit } from '@app/routes/hooks/rate-limit';

export enum CommentTarget {
  Episode = 1,
  Character = 2,
  Person = 3,
  Index = 4,
  Blog = 5,
  Timeline = 6,
}

const commentTables = {
  [CommentTarget.Episode]: schema.chiiEpComments,
  [CommentTarget.Character]: schema.chiiCrtComments,
  [CommentTarget.Person]: schema.chiiPrsnComments,
  [CommentTarget.Index]: schema.chiiIndexComments,
  [CommentTarget.Blog]: schema.chiiBlogComments,
  [CommentTarget.Timeline]: schema.chiiTimelineComments,
};

export class Comment {
  private readonly target: CommentTarget;
  private readonly table: (typeof commentTables)[CommentTarget];

  constructor(target: CommentTarget) {
    this.target = target;
    this.table = commentTables[target];
  }

  async getAll(mainID: number) {
    const data = await db.select().from(this.table).where(op.eq(this.table.mid, mainID));
    const uids = data.map((v) => v.uid);
    const users = await fetcher.fetchSlimUsersByIDs(uids);

    const comments: res.IComment[] = [];
    const replies: Record<number, res.ICommentBase[]> = {};

    let allReactions: Record<number, res.IReaction[]> = {};
    if (this.target === CommentTarget.Episode) {
      allReactions = await fetchTopicReactions(mainID, LikeType.EpReply);
    }

    for (const d of data) {
      const user = users[d.uid];
      const comment: res.ICommentBase = {
        id: d.id,
        mainID: d.mid,
        creatorID: d.uid,
        relatedID: d.related,
        content: d.content,
        createdAt: d.createdAt,
        state: d.state,
        reactions: allReactions[d.id],
      };
      if (d.related === 0) {
        comments.push({ ...comment, replies: [], user });
      } else {
        const rs = replies[d.related] ?? [];
        rs.push({ ...comment, user });
        replies[d.related] = rs;
      }
    }
    for (const comment of comments) {
      comment.replies = replies[comment.id] ?? [];
    }

    return comments;
  }

  async create(
    auth: Readonly<IAuth>,
    mainID: number,
    content: string,
    replyTo: number,
  ): Promise<{ id: number }> {
    if (!Dam.allCharacterPrintable(content)) {
      throw new BadRequestError('text contains invalid invisible character');
    }
    if (auth.permission.ban_post) {
      throw new NotAllowedError('create comment');
    }

    if (replyTo !== 0) {
      const [parent] = await db
        .select({ id: this.table.id, state: this.table.state })
        .from(this.table)
        .where(op.eq(this.table.id, replyTo))
        .limit(1);
      if (!parent) {
        throw new NotFoundError(`parent comment id ${replyTo}`);
      }
      if (parent.state !== CommentState.Normal) {
        throw new NotAllowedError(`reply to a abnormal state comment`);
      }
    }

    await rateLimit(LimitAction.Comment, auth.userID);

    const reply: typeof this.table.$inferInsert = {
      mid: mainID,
      uid: auth.userID,
      related: replyTo,
      content,
      createdAt: DateTime.now().toUnixInteger(),
      state: CommentState.Normal,
    };
    const [result] = await db.insert(this.table).values(reply);
    return { id: result.insertId };
  }

  async update(auth: Readonly<IAuth>, commentID: number, content: string) {
    const [current] = await db
      .select()
      .from(this.table)
      .where(op.eq(this.table.id, commentID))
      .limit(1);
    if (!current) {
      throw new NotFoundError(`comment id ${commentID}`);
    }
    if (current.uid !== auth.userID) {
      throw new NotAllowedError('edit a comment which is not yours');
    }
    if (current.state !== CommentState.Normal) {
      throw new NotAllowedError(`edit to a abnormal state comment`);
    }

    const [reply] = await db
      .select({ id: this.table.id })
      .from(this.table)
      .where(op.and(op.eq(this.table.mid, current.mid), op.eq(this.table.related, current.id)))
      .limit(1);
    if (reply) {
      throw new NotAllowedError('cannot edit a comment with replies');
    }

    await db.update(this.table).set({ content }).where(op.eq(this.table.id, commentID));

    return {};
  }

  async delete(auth: Readonly<IAuth>, commentID: number) {
    const [comment] = await db
      .select()
      .from(this.table)
      .where(op.eq(this.table.id, commentID))
      .limit(1);
    if (!comment) {
      throw new NotFoundError(`comment id ${commentID}`);
    }
    if (comment.uid !== auth.userID) {
      throw new NotAllowedError('delete a comment which is not yours');
    }
    if (comment.state !== CommentState.Normal) {
      throw new NotAllowedError('delete a abnormal state comment');
    }

    // 其它表暂时没有 state 字段
    switch (this.table) {
      case schema.chiiEpComments:
      case schema.chiiCrtComments:
      case schema.chiiPrsnComments: {
        await db
          .update(this.table)
          .set({ state: CommentState.UserDelete })
          .where(op.eq(this.table.id, commentID))
          .limit(1);
        break;
      }
      default: {
        await db.delete(this.table).where(op.eq(this.table.id, commentID)).limit(1);
      }
    }

    return {};
  }
}
