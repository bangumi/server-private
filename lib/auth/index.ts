import * as crypto from 'node:crypto';

import { createError } from '@fastify/error';
import { compare } from '@node-rs/bcrypt';
import { DateTime } from 'luxon';

import { TypedCache } from '@app/lib/cache';
import type { IUser, Permission } from '@app/lib/orm';
import { AccessTokenRepo, fetchPermission, fetchUserX } from '@app/lib/orm';
import * as orm from '@app/lib/orm';
import { intval } from '@app/lib/utils';
import NodeCache from '@app/vendor/node-cache.ts';

const tokenPrefix = 'Bearer ';
export const NeedLoginError = createError<[string]>(
  'NEED_LOGIN',
  'you need to login before %s',
  401,
);
export const NotAllowedError = createError<[string]>(
  'NOT_ALLOWED',
  `you don't have permission to %s`,
  401,
);
const HeaderInvalidError = createError<[string]>('AUTHORIZATION_INVALID', '%s', 401);

const TokenNotValidError = createError<[]>(
  'TOKEN_INVALID',
  "can't find access token or it has been expired",
  401,
);

export const enum UserGroup {
  Unknown = 0,
  Admin = 1,
  BangumiAdmin = 2,
  WindowAdmin = 3,
  Quite = 4,
  Banned = 5,
  // 不太清楚具体是什么
  _6 = 6,
  // 不太清楚具体是什么
  _7 = 7,
  CharacterAdmin = 8,
  WikiAdmin = 9,
  Normal = 10,
  WikiEditor = 11,
}

export interface IAuth {
  userID: number;
  login: boolean;
  allowNsfw: boolean;
  permission: Readonly<Permission>;
  /** Unix time seconds */
  regTime: number;
  groupID: UserGroup;
}

export async function byHeader(key: string | string[] | undefined): Promise<IAuth | null> {
  if (!key) {
    return emptyAuth();
  }

  if (Array.isArray(key)) {
    throw new HeaderInvalidError("can't providing multiple access token");
  }

  if (!key.startsWith(tokenPrefix)) {
    throw new HeaderInvalidError('authorization header should have "Bearer ${TOKEN}" format');
  }

  const token = key.slice(tokenPrefix.length);
  if (!token) {
    throw new HeaderInvalidError('authorization header missing token');
  }

  return await byToken(token);
}

const tokenAuthCache = TypedCache<string, IUser>((token) => `auth:token:${token}`);

export async function byToken(accessToken: string): Promise<IAuth | null> {
  const cached = await tokenAuthCache.get(accessToken);
  if (cached) {
    return await userToAuth(cached);
  }

  const token = await AccessTokenRepo.findOne({
    where: { accessToken, expires: orm.Gt(new Date()) },
  });

  if (!token) {
    throw new TokenNotValidError();
  }

  if (!token.userId) {
    throw new Error('access token without user id');
  }

  const u = await fetchUserX(intval(token.userId));

  await tokenAuthCache.set(accessToken, u, 60 * 60 * 24);

  return await userToAuth(u);
}

const userCache = TypedCache<number, IUser>((userID) => `auth:user:${userID}`);

export async function byUserID(userID: number): Promise<IAuth> {
  const cached = await userCache.get(userID);
  if (cached) {
    return await userToAuth(cached);
  }

  const u = await fetchUserX(userID);

  await userCache.set(userID, u, 60 * 60);

  return await userToAuth(u);
}

const permissionCache = new NodeCache({ stdTTL: 60 * 10 });

async function getPermission(userGroup?: number): Promise<Readonly<Permission>> {
  if (!userGroup) {
    return Object.freeze({});
  }

  const cached = permissionCache.get(userGroup);
  if (cached) {
    return cached;
  }

  const p = await fetchPermission(userGroup);

  permissionCache.set(userGroup, p);

  return p;
}

export function emptyAuth(): IAuth {
  return {
    userID: 0,
    login: false,
    permission: {},
    allowNsfw: false,
    regTime: 0,
    groupID: 0,
  };
}

async function userToAuth(user: IUser): Promise<IAuth> {
  return {
    userID: user.id,
    login: true,
    permission: await getPermission(user.groupID),
    allowNsfw: user.regTime - DateTime.now().toUnixInteger() <= 60 * 60 * 24 * 90,
    regTime: user.regTime,
    groupID: user.groupID,
  };
}

function processPassword(s: string): string {
  return crypto.createHash('md5').update(s).digest('hex');
}

export async function comparePassword(hashed: string, input: string): Promise<boolean> {
  return compare(processPassword(input), hashed);
}
