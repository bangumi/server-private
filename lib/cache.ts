import { redisPrefix } from '@app/lib/config';
import redis from '@app/lib/redis';

interface TypeSafeCacheUtil<Key, Value> {
  set(key: Key, value: Value, ttl?: number): Promise<void>;

  get(key: Key): Promise<Value | null>;

  del(key: Key): Promise<void>;

  cached(key: Key, getter: () => Promise<Value | null>, ttl?: number): Promise<Value | null>;
}

export function TypedCache<K, V>(
  toRedisKey: (v: K) => string,
  defaultTTL = 60,
): TypeSafeCacheUtil<K, V> {
  const buildKey = (key: K) => `${redisPrefix}:${toRedisKey(key)}`;

  return {
    async set(key: K, value, ttl = defaultTTL) {
      await redis.set(buildKey(key), JSON.stringify(value), 'EX', ttl);
    },
    async get(key: K): Promise<V | null> {
      const cached = await redis.get(buildKey(key));
      if (cached) {
        return JSON.parse(cached) as V;
      }

      return null;
    },
    async del(key: K): Promise<void> {
      await redis.del(buildKey(key));
    },

    async cached(key: K, getter: () => Promise<V | null>, ttl?: number): Promise<V | null> {
      const cached = await this.get(key);
      if (cached) {
        return cached;
      }

      const data = await getter();
      if (!data) {
        return null;
      }

      await this.set(key, data, ttl);

      return data;
    },
  };
}
