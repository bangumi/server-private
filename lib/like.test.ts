import { expect, test } from 'vitest';

import { fetchGroupTopic } from '@app/lib/like';

test('group topic reactions', async () => {
  await expect(fetchGroupTopic(379821, 406430)).resolves.toMatchSnapshot();
});
