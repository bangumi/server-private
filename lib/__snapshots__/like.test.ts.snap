// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`group topic reactions 1`] = `
Object {
  "2288668": Array [
    Object {
      "selected": false,
      "total": 2,
      "value": 54,
    },
    Object {
      "selected": false,
      "total": 2,
      "value": 68,
    },
    Object {
      "selected": true,
      "total": 1,
      "value": 92,
    },
    Object {
      "selected": false,
      "total": 1,
      "value": 122,
    },
    Object {
      "selected": false,
      "total": 1,
      "value": 128,
    },
    Object {
      "selected": false,
      "total": 2,
      "value": 132,
    },
    Object {
      "selected": false,
      "total": 1,
      "value": 137,
    },
  ],
  "2288669": Array [
    Object {
      "selected": false,
      "total": 1,
      "value": 68,
    },
    Object {
      "selected": false,
      "total": 1,
      "value": 85,
    },
    Object {
      "selected": true,
      "total": 1,
      "value": 90,
    },
    Object {
      "selected": false,
      "total": 1,
      "value": 92,
    },
    Object {
      "selected": false,
      "total": 1,
      "value": 122,
    },
    Object {
      "selected": false,
      "total": 1,
      "value": 132,
    },
  ],
  "2288670": Array [
    Object {
      "selected": true,
      "total": 1,
      "value": 60,
    },
    Object {
      "selected": false,
      "total": 2,
      "value": 62,
    },
    Object {
      "selected": false,
      "total": 1,
      "value": 122,
    },
    Object {
      "selected": false,
      "total": 1,
      "value": 128,
    },
    Object {
      "selected": false,
      "total": 3,
      "value": 137,
    },
  ],
}
`;
